### ISSUE TITLE
Showing status of service via systemctl is slow (>10s)  if disk journal is used

### ISSUE THREAD

**Issue by XANi** [systemd/systemd#2460]
With big (4GB, few months of logs) on-disk journal, `systemctl status service` becomes very slow

```
 (!) [13:37:30]:/var/log/journal☠ time  systemctl status nginx
* nginx.service - A high performance web server and a reverse proxy server
   Loaded: loaded (/lib/systemd/system/nginx.service; enabled; vendor preset: enabled)
   Active: active (running) since Thu 2016-01-28 08:25:28 CET; 5h 12min ago
 Main PID: 3414 (nginx)
   CGroup: /system.slice/nginx.service
           |-3414 nginx: master process /usr/sbin/nginx -g daemon on; master_process on;
           |-3415 nginx: worker process
           |-3416 nginx: worker process
           |-3417 nginx: worker process
           `-3418 nginx: worker process

Jan 28 08:25:12 ghroth systemd[1]: Starting A high performance web server and a reverse proxy server...
Jan 28 08:25:28 ghroth systemd[1]: Started A high performance web server and a reverse proxy server.

real    0m12.505s
user    0m0.016s
sys 0m0.056s
 (!) [13:35:14]:/var/log/journal☠ du -h --max=1
4,2G    ./46feef66e59512fcd99c7ddc00000108
4,2G    .
 (!) [13:40:53]:/var/log/journal☠ strace systemctl status nginx 2>&1 |grep open |grep /var/log |wc -l
88
```

it is of course faster after it gets to cache... for that service, querying other one is still slow.

Dunno what would be right way to do it.. but opening ~80 log files to just display service status seems a bit excessive

**Comment by poettering** [comment]
is this on hdd or ssd? 

But yeah, we scale badly if we have too many individual files to combine. It's O(n) with each file we get...

**Comment by XANi** [comment]
This one was on HDD, but now that I've looked into it it can be a bit (~0.5s, HDD swap, 300MB of logs on tmpfs) slow even with tmpfs if part that was loaded happened to be swapped out (I was testing on system with 2 weeks of uptime)

Shouldn't there be some kind of index on journal files ? Or at the very least pointer in service entry to last log file that have relevant log entries .

**Comment by davispuh** [comment]
I think some kind of journal indexing is required because it's unbearably slow.

Right now I've 5411 (43 GiB) journal files and

``` bash
$ time -p journalctl -b --no-pager > /dev/null
real 13.61
user 13.37
sys 0.22
```

it takes 13 seconds to just check current boot log while it's already cached in RAM.

When it's not cached

```
# echo 3 > /proc/sys/vm/drop_caches
$ time -p journalctl -b --no-pager > /dev/null
real 69.29
user 13.51
sys 12.34
```

this is on 2x 3TB HDD with RAID1 btrfs.

**Comment by XANi** [comment]
It is laggy even when it is on tmpfs and machine runs long enough to swap it out.

Why journald doesn't just use SQLite for storage ? It would be faster and other apps could actually use logfiles for something useful and have good query language instead of relying on a bunch of options in journalctl

**Comment by XANi** [comment]
It is still slow as hell:

    time systemctl status kdm|cat
    * kdm.service - LSB: X display manager for KDE
       Loaded: loaded (/etc/init.d/kdm; generated; vendor preset: enabled)
       Active: active (exited) since Wed 2017-06-21 12:03:26 CEST; 1h 42min ago
         Docs: man:systemd-sysv-generator(8)
        Tasks: 0 (limit: 4915)
       CGroup: /system.slice/kdm.service
    
    Jun 21 12:03:25 ghroth systemd[1]: Starting LSB: X display manager for KDE...
    Jun 21 12:03:26 ghroth systemd[1]: Started LSB: X display manager for KDE.
    
    real	0m2.873s
    user	0m0.008s
    sys	0m0.016s

and it opens over hundred (on system that was up 2 hours)

    strace systemctl status kdm 2>&1 |grep open |wc -l
    133

**Comment by vcaputo** [comment]
@XANi @davispuh is there any chance you could run your slow cases under `valgrind --tool=callgrind` and supply us with the the output?  I've seen some pathological cases where `journalctl` is spending exorbitant (~50% of its execution) amounts of CPU time in the siphash24 code when the simple context cache in the mmap-cache is experiencing high miss rates, I'm curious if that's also something you guys are observing here.

**Comment by davispuh** [comment]
with systemd 233.75-3 on ArchLinux

[callgrind.out.systemctl-status-sshd.gz](https://github.com/systemd/systemd/files/1133065/callgrind.out.systemctl-status-sshd.gz)
[callgrind.out.journalctl-b--no-pager.gz](https://github.com/systemd/systemd/files/1133072/callgrind.out.journalctl-b--no-pager.gz)

**Comment by vcaputo** [comment]
@davispuh Thank you for quickly providing the profiles.  The `callgrind.out.systemctl-status-sshd.gz` profile shows `mmap_cache_get()` somewhat prominently with the hashmap maintenance being a significant part of that.

It's not a panacea, but https://github.com/systemd/systemd/pull/6307 may improve the runtime of `systemctl status sshd` for you.  Would you be up for some testing?  Some `time systemctl status sshd --no-pager` comparisons before and after would be great.

For anybody reading, the `kcachegrind` utility works quite well for visualizing callgrind.out files.

**Comment by davispuh** [comment]
Emm, I can't test this anymore since after I compiled and reinstalled systemd it reset my `/etc/systemd/journald.conf` settings and so systemd deleted all old journals and it's fast now again. Basically it's slow only when you've a lot of journal files in `/var/log/journal/`

**Comment by davispuh** [comment]
With just complied from fdb63434844f53a14073000931263c7f81d82bab

When files aren't cached it's really unusably slow
```
$ time systemctl status sshd --no-pager
0.28user 34.82system 4:34.32elapsed 12%CPU (0avgtext+0avgdata 1355416maxresident)k
3242224inputs+0outputs (10002major+17658minor)pagefaults 0swaps
```

2nd time when it's cached it's quick
```
$ time systemctl status sshd --no-pager
0.09user 0.33system 0:00.43elapsed 99%CPU (0avgtext+0avgdata 1303540maxresident)k
0inputs+0outputs (0major+23705minor)pagefaults 0swaps
```

`callgrind` when it's not cached and when it is

[callgrind.out.systemctl-status-sshd_nocache.gz](https://github.com/systemd/systemd/files/1361087/callgrind.out.systemctl-status-sshd_nocache.gz)
[callgrind.out.systemctl-status-sshd_cached.gz](https://github.com/systemd/systemd/files/1361088/callgrind.out.systemctl-status-sshd_cached.gz)

I've 7542 `*.journal` files in `/var/log/journal/<ID>` which is on BTRFS RAID1 partition (2x 3TB HDD)

Basically to improve performance need to do less disk reading. Like use some kind of indexing or something like that.

**Comment by amishxda** [comment]
I have 240 system log files and 860 user log files.

systemctl status OR journalctl -f take 2-4 minutes just to display logs. (HDD drive)

I have added this in: /usr/lib/systemd/journald.conf.d/90-custom.conf

```
[Journal]
SystemMaxUse=100G
SystemMaxFileSize=200M
SystemMaxFiles=1100
MaxFileSec=1week
MaxRetentionSec=3year
```

Systemd generates 2 to 3 system journal files everyday each of about 150994944 bytes size.

Why doesnt journalctl -f (or systemctl) check only latest / current journal?

How do I make it efficient and fast? I need to preserve logs for long duration.

In most cases most people only have to check recent logs only.

May be some feature to have automatic archival of logs in different directory (/var/log/journal/ID-DIRECTORY/archive) and current logs (say past 3-7 days) kept in /var/log/journal/ID-DIRECTORY?

This will speed up journalctl and systemctl status a lot. Anyone want to check archived logs can use --directory option of journalctl

**Comment by amishxda** [comment]
journalctl now has --file parameter. I am now able to use it to search faster.

while:
`journalctl -f` took ages

`journalctl --file /var/log/journal/*/system.journal -f` takes just 1-2 seconds.

Similarly can we make `systemctl status` use ONLY system.journal by default? Because in most cases administrator check status only after `systemctl (re)start` or when something unexpected happens (which is also likely to be logged in system.journal unless it rotated just recently)

This will drastically speed it up.

If admin wants older status he can supply `--lines=N` in which case systemctl will scan through older journals too. (may be in reverse order) OR admin can use `journalctl -n N -u service` instead

PS: I have no idea how data is stored in journal.

@poettering do u want me to create RFE for this?

**Comment by Gunni** [comment]
Yes this is a problem for servers storing their logs, my biggest problem is the centralized log server, i receive logs from network equipment using rsyslog, but it uses omjournal to pipe them directly into journal. It works fine to begin with but then degrades quickly.

Maybe journal files could be made to contain specific timespans, and only get loaded if requested, i use stuff like `--since` and `--until` a lot they are affected by this, but if the file method were used, journalctl could find them in fractions of seconds, `-f` should just be instant really, find the last 10 entries, and only search backwards in time if nothing happened maybe with a cutoff.

**Comment by XANi** [comment]
@amishxda IMO `status` by default should just... not have to touch on-disk logs, ever. Storing few last lines of log in memory per service wouldn't be that memory hungry, wouldn't trash OS disk cache on `status` and it would return something more useful than "welp, logs were rotated, too bad" that it currently does when server runs for some time.

It should only fetch more log lines from journalctl if explictly requested by user


@Gunni I don't think using systemd for centralized log server is intended, or good idea in the first place. ELK stack is much more useful for it, jankiness aside Logstash allows to do a lot of nice stuff, for example we use it to split iptables logs into fields (srcip/dstip/port etc) before putting in ES. And ES just works better for search

**Comment by Gunni** [comment]
@XANi I just expected it to be a supported use case since `systemd-journal-remote` is a thing but if this is the expected performance, ofcourse. But if it is improved, maybe add a better index or a database backend to journal then i'd be able to use it.

About the tools you mentioned, setting up all that stuff sounds like much more work, especially since we like to be able to watch the logs live `journalctl SRC=net -f` (maybe with a grep), the second they arrive, my expirience with what you mentioned has been long delay from event to user. But i'll look into it.

**Comment by XANi** [comment]
@Gunni I wish journald would just use sqlite instead of its current half-assed binary db. I feel like currently it is just trying to reimplement that but badly. And there is a plenty of tools for querying sqlite already.

ELK stack is definitely more effort to put in but in exchange it has a ton of nice features to look at, we for example made logstash do geoIP lookup on any IP that's not private so each firewall log gets that info added.

Querying is also very nice as you can make a queries on fields directly instead of text strings

**Comment by otisg** [comment]
I found out about this issue via https://www.reddit.com/r/linuxadmin/comments/gdfi4t/how_do_you_look_at_journald_logs/ .  Is this still a problem?

**Comment by XANi** [comment]
@otisg as of systemd 241 (that's just version I have on machine that actually keeps logs on disk), it is most definitely still a problem (all timing tests done right after dropping cache, ~800MB in journal):
```
$ time systemctl status collectd  2>&1 >/dev/null
real	0m5,921s
user	0m0,004s
sys	0m0,180s

$ strace -f    systemctl status collectd  2>&1 |grep open |grep /var/log/journal |wc -l
721
```
now the fun part :
```
find /var/log/journal |wc -l
```
Yes, you are reading this right, getting a last few lines of a currently running service opens **every single fucking entry in the journal dir**

Now if I do just
```
$ time grep -R collectd /var/log/*log >/dev/null

real	0m0,370s
user	0m0,010s
sys	0m0,009s
```
This shit manages to be order of magnitude slower than "just bruteforce grepping last logrotate's worth of text logs. This is amount of retardness that's fucking mind-boggling. The sheer fact developers decided to go with binary format yet not bother by even introducing time or service-based sharding/indexing and just bruteforce every existing file is just insane.

It is like someone, instead of considering reasonable options like, dunno:

* naming log files by hash of the service name so it never needs to open more than 1/n files
* naming log files by service name (or service group) so there is only 1 file needed to open, ever for recent stuff and more is needed only for history (you know, like logrotate + syslog combo ends up being)
* using actual well tested database like SQLite and just create few bigger files with index/TOC

They decided one evening "you know, I always wanted to make a binary logging format", then got bored after few weeks and never touched it again.

**Comment by otisg** [comment]
Ouch.  I had not realized things were so slow.  So I'm amazed why **so many** people at https://www.reddit.com/r/linuxadmin/comments/gdfi4t/how_do_you_look_at_journald_logs/ said they consume journal logs via journalctl.  Are they all OK with slowness?!?  Why don't more people get their logs out of journal, centralize them in an on-prem or SaaS service that is faster?

Anyhow, I see some systemd developers here.  I wonder if they plan on integrating something like https://github.com/tantivy-search/tantivy ...

**Comment by PAStheLoD** [comment]
The first invocation is slow. It probably goes through the journal files and checks them. Then things are fast for a while.

journald is not great for log management, and it's simply not fit for any kind of centralized log management at scale. But it's very likely not a goal of the systemd project to handle even that too. The journal is a necessity, just as pid1, udev, and network setup (for remote filesystems) to manage a Linux system and its services _reliably_.

That said it's entirely likely that with a few quick and dirty optimizations this could be worked around. (Eg. if journal files are not in cache, don't wait for them when showing status; allow streaming the journal without looking up the last 10 lines, persisting some structures to speed up journal operations, enabling unverified journal reads by default, etc.)

**Comment by vcaputo** [comment]
@XANi Are you still suffering from these very slow uncached `systemctl status collectd` invocations?

If so, curious if you'd be willing to try out a build with https://github.com/systemd/systemd/pull/17459 applied and report back the results, thanks.

**Comment by XANi** [comment]
@vcaputo Correct me if I'm wrong but that patch only searches in current boot id, so if system is running for a long time it wouldn't change anything ?

I have encountered the problem on server machines in the first place (and on personal NAS) so in almost every case the current boot ID is only one in the logs. It certainly would help on desktop but that's not where I hit the problem in the first place (also AFAIK most desktop distros don't have /var/log/journal by default so journalctl doesn't log to HDD in the first place).

The other problem is that current implementation makes it *really* easy for one service to swamp the logs  to the point you lose any other service's logs. It is especially apparent for services that don't emit much logs, and even tho there is zero *actual* logs for the service (as they get rotated out) it still takes about the same amount of time as for any other service.

For reference I have it happening on machine with just 1GB of journal and last few days of logs there.

**Comment by vcaputo** [comment]
@XANi Yes you're right, if all the journals are from the same boot the early exit never occurs.

FTR it already matches the current boot id, the patch just moves the order around.  The change assumes there are multiple boots represented in the archived journals.

**Comment by amstan** [comment]
Sorry if I'm stating the obvious, but I might not be understanding this right.
While I might understand that reading multi GB of systemd binary logs will take a little bit of time, and stuff that's cached will probably come instantly.

But this is a common pattern for me and I still have to wait, even though the logs were produced mere seconds ago.

```
alex@hypertriangle:~% date
Sun Jan  3 03:18:30 EST 2021
alex@hypertriangle:~% time sudo systemctl restart ntpd  
0.21s real  0.01s user  0.02s system  14% cpu  6kB mem $ sudo systemctl restart ntpd
alex@hypertriangle:~% time sudo systemctl status ntpd 
● ntpd.service - Network Time Service
     Loaded: loaded (/usr/lib/systemd/system/ntpd.service; enabled; vendor preset: disabled)
     Active: active (running) since Sun 2021-01-03 03:18:34 EST; 5s ago
    Process: 152305 ExecStart=/usr/bin/ntpd -g -u ntp:ntp (code=exited, status=0/SUCCESS)
   Main PID: 152308 (ntpd)
      Tasks: 2 (limit: 4621)
     Memory: 1.4M
     CGroup: /system.slice/ntpd.service
             └─152308 /usr/bin/ntpd -g -u ntp:ntp

Jan 03 03:18:34 hypertriangle.com ntpd[152308]: Listen and drop on 0 v6wildcard [::]:123
.....snip....
Jan 03 03:18:34 hypertriangle.com systemd[1]: Started Network Time Service.
5.58s real  0.01s user  0.36s system  6% cpu  41kB mem $ sudo systemctl status ntpd
```
5.6 seconds to output a few lines that were generated a few seconds ago by the restart call.

True, another call immediatelly after is instant, but by then it's too late.
```
alex@hypertriangle:~% time sudo systemctl status ntpd
● ntpd.service - Network Time Service
     Loaded: loaded (/usr/lib/systemd/system/ntpd.service; enabled; vendor preset: disabled)
     Active: active (running) since Sun 2021-01-03 03:18:34 EST; 41s ago
    Process: 152305 ExecStart=/usr/bin/ntpd -g -u ntp:ntp (code=exited, status=0/SUCCESS)
   Main PID: 152308 (ntpd)
      Tasks: 2 (limit: 4621)
     Memory: 2.2M
     CGroup: /system.slice/ntpd.service
             └─152308 /usr/bin/ntpd -g -u ntp:ntp

Jan 03 03:18:34 hypertriangle.com ntpd[152308]: Listen and drop on 0 v6wildcard [::]:123
.....snip....
Jan 03 03:18:34 hypertriangle.com systemd[1]: Started Network Time Service.
0.42s real  0.01s user  0.04s system  13% cpu  41kB mem $ sudo systemctl status ntpd
alex@hypertriangle:~% date
Sun Jan  3 03:18:53 EST 2021
```

This is particularly annoying when you don't know why some service didn't restart properly and checking up on it essentially punishes you with the delay.

**Comment by vcaputo** [comment]
If re-running  `systemctl status` returns immediately, it's purely the cost of reading the journals from storage in the uncached case that's taking so long.

Unfortunately, substantially improving the uncached read performance is probably going to require changes to the journal file format.  There might be some gains to be found in filesystem tuning, like trying to tune /var/log/journal's filesystem for granular mmap-oriented pagefault-based reads in random-ish access patterns.  I haven't personally explored that avenue, but it might be worth exploring.

One reason uncached performance sucks is most journal accesses are far smaller than a 4KiB page, especially in search-like operations as `systemctl status` performs. But the accesses are fulfilled entirely via address space mappings setup via mmap(), so even when the code only tries to read 8 bytes to see some object's offset or seqnum, at least 4KiB get faulted in to fulfill the access; that's the granularity of memory mappings.  If the accesses were performed with explicit read() calls, maybe it could be whittled down to 512-byte loads, assuming storage having 512-byte sector sizes on complementary configured filesystem, but then syscall overhead would be painful, an io_uring implementation might be interesting on that front.

Due to the substantial lack of clustering in the format to try ensure faulted-in pages bring in operation-relevant information in the baggage of the 4KiB+ page fault of something like the acces of an 8-byte offset or seqnum, most of the baggage often isn't even immediately useful (though it's at least now warm in the page cache for subsequent journal operations).  Due to the combination of the tiny objects and the inherent interleaving of types of objects in a largely appended-as-arrived layout, faulted in pages tend to not contain stuff immediately needed by the faulting operation.

It's not that the format lacks indexing as many have complained, as indicated by the cached performance; the results are immediate when not waiting for page faults from backing store.  It tends to more often be a sort of read-amplification effect of the append-mostly tiny objects interleaved by type accessed in units of pages that just ends up reading a whole lot more crap than is relevant to the operation at hand, when uncached.

**Comment by HLFH** [comment]
Frankly, the search on my IMAP server (Dovecot) search was broken (too slow), and when I started to use a [full-text search indexing plugin that uses ElasticSearch as a backend](https://github.com/filiphanes/fts-elastic), it was so fast! Cannot systemctl & journalctl use ElasticSearch indexing capabilities in conjunction to the redesign of the journal file format as advised by @vcaputo?

**Comment by vcaputo** [comment]
In case it's unclear: `systemd-journald` is *not* anything remotely resembling a persistent database daemon, it's not really comparable to database "backends".

Whenever `journalctl` or `systemctl status` processes must access journal data, they do so in total isolation by faulting journal file pages into their own address space directly from the filesystem.  This is sort of like a cold read-only database startup on behalf of every journal-accessing process.  There is no persistent daemon servicing these clients, pre-warming the journal data and/or trying to keep prioritized parts of it memory resident.

I'm not the designer or decision maker of these things, but it seems clear based on what I've gleaned working with the code that this is all deliberate in an attempt to keep the overall constant journald footprint foisted on *every* systemd host to a minimum.

I imagine if we added a persistent database daemon to the architecture's read-side, the handful of people complaining about cold cache performance will be replaced with people complaining about some journal database daemon wasting resources on things which rarely occur.

Frankly, if people are generally satisfied with the *cached* performance of these commands, and the featureset offered, you could just schedule a `systemctl status $foo` or something to occur regularly as a cache warmer in lieu of a persistent database daemon serving readers and call it Good Enough.

For more elaborate search features, can't people already bolt on more sophisticated indexers entirely external to journald as-is?

**Comment by davispuh** [comment]
> Frankly, if people are generally satisfied with the cached performance of these commands, and the featureset offered, you could just schedule a systemctl status $foo or something (journalctl --verify? cat /var/log/journal/*/* > /dev/null?) to occur regularly as a cache warmer in lieu of a persistent database daemon serving readers and call it Good Enough.

It's not really practical if you keep logs for a really long time. Imagine 2 years worth of logs, it's pointless to have them cached. When you need only last month usually. But maybe once a year you want to look more in back.

> For more elaborate search features, can't people already bolt on more sophisticated indexers entirely external to journald as-is?

This sounds like best solution. But I'm not aware of anything like that. Basically it looks like systemd journal can't handle our needs so we need something better but it would be really great if it could be plugged in journald so that we don't need to learn new tools. eg. use `systemctl status`, `journalctl` independently if there's a separate indexing/storage backend.

**Comment by marcor** [comment]
For the record this is the reason I first considered switching to distros not using systemd, and now that I understand and see how the root cause is still here after many years, I'm reassured that it was a good decision after all.

**Comment by XANi** [comment]
> I'm not the designer or decision maker of these things, but it seems clear based on what I've gleaned working with the code that this is all deliberate in an attempt to keep the overall constant journald footprint foisted on _every_ systemd host to a minimum.

Well, except that the way it does it trashes cache. Cached performance is attained only in condition where a ton of useless logs are loaded into memory only to find last few lines

> I imagine if we added a persistent database daemon to the architecture's read-side, the handful of people complaining about cold cache performance will be replaced with people complaining about some journal database daemon wasting resources 

I imagine slapping an index that just stored "latest journal file service is in" would solve ~99% issues with getting status here.

I imagine if it had semi-sensible system of splitting logs between filenames (via name hash or whatever) would cut it down by order of magnitude or two, altho probably generate more random IO so probably not a tradeoff worth taking.

Current performance is worse than tailing a logfile.

> on things which rarely occur.

Nonsense. Anything under configuration management and proper monitoring will have that run often, from once or twice per hour to multiple times per minute. So at best wasting cache on some files you don't even pick any data from.

**Comment by mbiebl** [comment]
When you run `systemctl status` from your monitoring system, you are most likely not really interested in the last few journal lines and you could use `systemctl status --lines=0` to disable journal output.

**Comment by XANi** [comment]
@mbiebl if service is working, sure. If service isn't, I'm VERY interested in last few lines. Same for monitoring checks really, but I guess there it is always possible to do `systemctl is-active xyz || systemctl status` instead

But, if I'm checking in the first place, there is a good chance something is wrong with it so not that useful for CLI work.

**Comment by amstan** [comment]
> I imagine slapping an index that just stored "latest journal file service is in" would solve ~99% issues with getting status here.

Yes please, this is the solution. And I agree with the 99% number.

> Current performance is worse than tailing a logfile.

True.

**Comment by jhass** [comment]
Could systemctl show not flush the header and then block while fetching the log? Then one could just Ctrl-C it when not interested in the log anymore after checking the status.

**Comment by mildred** [comment]
Still having this problem on a HDD with journald. Using btrfs with nodatacow and frequent defragment. Journal files are few (63 files) and not fragmented, but this is unbareably slow. Read and writes are so slow than connecting via ssh to the machine takes several seconds just for the ssh daemon to be able to write to the journal I suppose.

**Comment by bombela** [comment]
`journalctl` pathologically slow. It is slow on NVME/SSD/HDD. It is slow whether the cache is hot or cold. I regularly come back to his issue, hoping one day it will be fixed.

It is so slow, that I find it faster to first dump all the logs into a file, and grepping through it thereafter.

```
$ time journalctl > /tmp/all.log

real    1m11.364s
user    0m52.299s
sys     0m6.540s

$ time wc -l /tmp/all.log
3659597 /tmp/all.log

real    0m0.152s
user    0m0.056s
sys     0m0.096s

$ time journalctl | grep sshd | wc -l

12944

real    0m53.973s
user    0m49.535s
sys     0m5.210s

$ time grep sshd /tmp/all.log | wc -l
12944

real    0m0.429s
user    0m0.332s
sys     0m0.100s
```

`/tmp` and `/var/log` are on the same filesystem.
```
stat -c '%d' /var/log/journal /tmp
64771
64771
```

Of course back in the days, we would have multiple log files and logrotate. So you could grep faster after narrowing down to the logfile.

So `journalctl` is 120x slower than a plain old log file. How can this be?

**Comment by taoky** [comment]
https://github.com/systemd/systemd/pull/29261 may help resolve this performance issue in journalctl. And on my laptop, `strace` shows that `journalctl` is just busy `newfstatat()`ing files (which seems to be the problem #29261 fixes):

```
> strace -f journalctl | grep gnome | wc -l
...
newfstatat(57, "", {st_mode=S_IFREG|0640, st_size=8388608, ...}, AT_EMPTY_PATH) = 0
newfstatat(9, "", {st_mode=S_IFREG|0640, st_size=5396720, ...}, AT_EMPTY_PATH) = 0
newfstatat(58, "", {st_mode=S_IFREG|0640, st_size=4614272, ...}, AT_EMPTY_PATH) = 0
newfstatat(58, "", {st_mode=S_IFREG|0640, st_size=4614272, ...}, AT_EMPTY_PATH) = 0
newfstatat(9, "", {st_mode=S_IFREG|0640, st_size=5396720, ...}, AT_EMPTY_PATH) = 0
newfstatat(58, "", {st_mode=S_IFREG|0640, st_size=4614272, ...}, AT_EMPTY_PATH) = 0
newfstatat(9, "", {st_mode=S_IFREG|0640, st_size=5396720, ...}, AT_EMPTY_PATH) = 0
newfstatat(59, "", {st_mode=S_IFREG|0640, st_size=8388608, ...}, AT_EMPTY_PATH) = 0
newfstatat(60, "", {st_mode=S_IFREG|0640, st_size=5304608, ...}, AT_EMPTY_PATH) = 0
newfstatat(60, "", {st_mode=S_IFREG|0640, st_size=5304608, ...}, AT_EMPTY_PATH) = 0
newfstatat(9, "", {st_mode=S_IFREG|0640, st_size=5396720, ...}, AT_EMPTY_PATH) = 0
newfstatat(60, "", {st_mode=S_IFREG|0640, st_size=5304608, ...}, AT_EMPTY_PATH) = 0
newfstatat(9, "", {st_mode=S_IFREG|0640, st_size=5396720, ...}, AT_EMPTY_PATH) = 0
newfstatat(61, "", {st_mode=S_IFREG|0640, st_size=4827904, ...}, AT_EMPTY_PATH) = 0
newfstatat(65, "", {st_mode=S_IFREG|0640, st_size=15603472, ...}, AT_EMPTY_PATH) = 0
newfstatat(9, "", {st_mode=S_IFREG|0640, st_size=5396720, ...}, AT_EMPTY_PATH) = 0
newfstatat(65, "", {st_mode=S_IFREG|0640, st_size=15603472, ...}, AT_EMPTY_PATH) = 0
newfstatat(9, "", {st_mode=S_IFREG|0640, st_size=5396720, ...}, AT_EMPTY_PATH) = 0
newfstatat(62, "", {st_mode=S_IFREG|0640, st_size=4575824, ...}, AT_EMPTY_PATH) = 0
newfstatat(63, "", {st_mode=S_IFREG|0640, st_size=16130824, ...}, AT_EMPTY_PATH) = 0
newfstatat(65, "", {st_mode=S_IFREG|0640, st_size=15603472, ...}, AT_EMPTY_PATH) = 0
newfstatat(9, "", {st_mode=S_IFREG|0640, st_size=5396720, ...}, AT_EMPTY_PATH) = 0
newfstatat(65, "", {st_mode=S_IFREG|0640, st_size=15603472, ...}, AT_EMPTY_PATH) = 0
newfstatat(9, "", {st_mode=S_IFREG|0640, st_size=5396720, ...}, AT_EMPTY_PATH) = 0
newfstatat(64, "", {st_mode=S_IFREG|0640, st_size=109441536, ...}, AT_EMPTY_PATH) = 0
newfstatat(65, "", {st_mode=S_IFREG|0640, st_size=15603472, ...}, AT_EMPTY_PATH) = 0
newfstatat(65, "", {st_mode=S_IFREG|0640, st_size=15603472, ...}, AT_EMPTY_PATH) = 0
newfstatat(9, "", {st_mode=S_IFREG|0640, st_size=5396720, ...}, AT_EMPTY_PATH) = 0
newfstatat(65, "", {st_mode=S_IFREG|0640, st_size=15603472, ...}, AT_EMPTY_PATH) = 0
newfstatat(9, "", {st_mode=S_IFREG|0640, st_size=5396720, ...}, AT_EMPTY_PATH) = 0
newfstatat(66, "", {st_mode=S_IFREG|0640, st_size=4870200, ...}, AT_EMPTY_PATH) = 0
newfstatat(68, "", {st_mode=S_IFREG|0640, st_size=4189672, ...}, AT_EMPTY_PATH) = 0
newfstatat(9, "", {st_mode=S_IFREG|0640, st_size=5396720, ...}, AT_EMPTY_PATH) = 0
newfstatat(68, "", {st_mode=S_IFREG|0640, st_size=4189672, ...}, AT_EMPTY_PATH) = 0
newfstatat(9, "", {st_mode=S_IFREG|0640, st_size=5396720, ...}, AT_EMPTY_PATH) = 0
newfstatat(67, "", {st_mode=S_IFREG|0640, st_size=89986696, ...}, AT_EMPTY_PATH) = 0
newfstatat(68, "", {st_mode=S_IFREG|0640, st_size=4189672, ...}, AT_EMPTY_PATH) = 0
newfstatat(68, "", ^Cstrace: Process 891023 detached
```

Performance on my laptop (`journalctl | grep gnome | wc -l`):

systemd 254 (Arch build) journalctl: 10+min CPU time and I just sigint it as I no longer have that kind of patience waiting for it.

#29261 (buildtype release) journalctl: 73.92s (user 63.08s + system 14.76s, looks like CPU is busy parsing binary journal and `write()`ing)

**Comment by davispuh** [comment]
Happy new year, nothing have changed here, still unbearably slow as usual even with latest systemd 255.

At this point I think I should just look into ingesting logs into some other backend, maybe start using [SigNoz](https://signoz.io/)

**Comment by rnhmjoj** [comment]
The `stat` problem is worked on: https://github.com/systemd/systemd/pull/30209.
I don't know if that's enough to solve the journal performance issues, though.

**Comment by vcaputo** [comment]
> The `stat` problem is worked on: #30209. I don't know if that's enough to solve the journal performance issues, though.

That code didn't even exist when this issue was created/discussed.  It was introduced in 2023 by https://github.com/systemd/systemd/commit/34af74946e8853411f18120007ebaca6549b2a52

So at best #30209 undoes a later perf regression introduced by adding that stat storm in 2023, it's not causal for what's going on here.

**Comment by davispuh** [comment]
> For the perspective:
> 
> ```
> % time systemctl status --no-pager systemd-networkd                                                                                                   
>  …                                                                                                                                                             
> real    0m17,491s                                                                                                                                             
> user    0m0,013s                                                                                                                                              
> sys     0m1,171s 
> ```
> 
> Tha't on quite recent systemd-254.7-1.fc39.x86_64, on nvme-cached, btrfs raid-1 on 2 spinning Seagate IronWolf Pro.

I wish it was only 20s, try to use journal when
```
# time systemctl status stuff
[...]
real    4m47,699s
user    0m0,421s
sys     2m24,206s
```

And this is after running it once before, first time it took over 8mins. And my whole `/var` is on NVME. Also my journal folder is only ~3.5 GiB so actually it's way slower than it was back then in 2015 when I originally commented so things have become only worse not better.

Looking at `iotop` I see `journalctl` reading 200 MiB/s on average so looks like it's reading over and over again same stuff what a joke.

Also `journalctl -b --no-pager -u stuff` took almost hour :joy:

**Comment by vcaputo** [comment]
https://github.com/systemd/systemd/pull/30209 isn't merged yet, v254 contains https://github.com/systemd/systemd/commit/34af74946e8853411f18120007ebaca6549b2a52 which introduced the fstats that hurt performance, so you're likely just experiencing the known regression at this point.

No significant changes have been made to the journal file format prioritizing improved reader/journalctl performance.

**Comment by XANi** [comment]
@vcaputo v254 is as slow as it was when I first reported that bug. Back then it took 12.5s to get the status out of 4GB of journals, now it takes 5s to do the same on 1GB (I have since then limited journal size to not make using it be so awful)

It's impossible for it to be fast if it doesn't keep track what was last log that given app used or had proper indexing (and at that point just use SQLite).

**Comment by vcaputo** [comment]
It's not like the journal lacks any acceleration structures, IIRC the main *expected* source of poor scaling is when there's a large number of journals participating in the query since that scales O(N) where N is the number of journals.  Per-journal the entry arrays are binary-searched and grow exponentially, so for any given journal it should be a relatively quick sequence iterating the start to the last entry array, no?  Perhaps I'm misremembering the details here.

I feel like there might be some value in doing some instrumenting and verifying the amount of journal data accessed is at least consistent with what's expected for the existing file format.  It's possible to accidentally have extraneous work performed while still producing correct results with this type of thing.

**Comment by XANi** [comment]
The edge case is where journal *doesn't* have any logs for a given service.

Imagine server where some of the services are writing logs regularly, while other might not output anything for days or weeks (say nginx which only emits logs on start and end, and otherwise writes its own log files elsewhere).

Once you get to the point where the "silent" service  gets rotated out of journald files and it takes long time to get even the last few lines it emitted.

I also have no idea why plain grepping the same journal files is faster...

**Comment by vcaputo** [comment]
> The edge case is where journal _doesn't_ have any logs for a given service.
> 
> Imagine server where some of the services are writing logs regularly, while other might not output anything for days or weeks (say nginx which only emits logs on start and end, and otherwise writes its own log files elsewhere).
> 
> Once you get to the point where the "silent" service gets rotated out of journald files and it takes long time to get even the last few lines it emitted.

Except that doesn't really make sense, based on my recollection of how this should work.  If that's what's happening in that edge case, I think it's a bug in the implementation.

The match of BOOT_ID is ANDed with all the other conditions.  Your edge case would be where all the other conditions fail to advance the location, and your description makes it sound like everything matching the BOOT_ID is then getting visited anyways.

My recollection is that this should just stop the matching from progressing further at all, short-circuiting without visiting all the entries for the boot id.

It could just be a silly bug to go fix.  Try running the `systemctl status` with SYSTEMD_LOG_LEVEL=debug and look for the journal filter line, it should make clear what the match conditions are.

**Comment by XANi** [comment]
> The match of _BOOT_ID is ANDed with all the other conditions. Your edge case would be where all the other conditions fail to advance the location, and your description makes it sound like everything matching the _BOOT_ID is then getting visited anyways.

It does look like it is. But even if it was not it would mean that just one spammy app would make any status/log request for any other app slow.

At the very least each service should store the pointer to last file it had its logs in so it could be short circuited instead of siting thru gigabytes of logs just to display its status. Or hell, slap a filename over



> It could just be a silly bug to go fix. Try running the systemctl status $silent.service with SYSTEMD_LOG_LEVEL=debug and look for the journal filter line, it should make clear what the match conditions are.

As you wish (systemd from Debian 12, 242.6):

```
# echo 3 > /proc/sys/vm/drop_caches
# time systemctl -l status nginx 2>&1 |ts >/tmp/log #save it to log with timestamps
real	0m3.819s
user	0m0.045s
sys	0m0.184s

# cat /tmp/log  |grep File |wc -l # get the number of "File ... added." lines
81

# find /var/log/journal -type f |wc -l # get number of files in the journal
81
``` 

Lo and behold, every single file visited.  Here is full log: [log.txt](https://github.com/systemd/systemd/files/14146794/log.txt)

Journal filter:

    Feb 03 01:02:17 Journal filter: (_BOOT_ID=a2e5ac42f2ea4e9f8f52080d4be63985 AND ((OBJECT_SYSTEMD_UNIT=nginx.service AND _UID=0) OR (UNIT=nginx.service AND _PID=1) OR (COREDUMP_UNIT=nginx.service AND _UID=0 AND MESSAGE_ID=fc2e22bc6ee647b6b90729ab34a250b1) OR _SYSTEMD_UNIT=nginx.service))

But let's test more. let's restart nginx so there are some fresh logs in journal:

```

# systemctl restart nginx
# echo 3 > /proc/sys/vm/drop_caches
# cat /tmp/log  |grep File  |wc -l
81
```
But hey, maybe that's not enough *new* lines and it have to get all the journal to return more than just "Starting nginx" ?

```
# for a in {1..2000} ; do systemctl restart nginx ; done # now my entire log is failed nginx restart requests because systemd didn't liked that many requests too fast, which is entirely understandable

# time systemctl -l status  nginx 2>&1 |ts >/tmp/log

real	0m4.006s
user	0m0.053s
sys	0m0.198s
# # it got slower, coz logs are now bigger

# cat /tmp/log  |grep File  |wc -l
81
# find /var/log/journal -type f |wc -l
81
```

Still scans every file even if it has recent logs available, despise command needing like 
10 last log lines to display.

**Comment by vcaputo** [comment]
Visiting every file is expected, they all participate in the matching process, that's by design.

What's unexpected is visiting every *entry* in every journal.  It shouldn't have to read in the entirety of every journal file to answer a `systemctl status $foo.service` style query AIUI.

I feel like we're talking past eachother and the issue's going in circles, but I don't have time right now to dig into the code and run this to ground.  My efforts historically mostly focused on the journald/writer side of things, not the consumer/reader side.

**Comment by XANi** [comment]
> What's unexpected is visiting every entry in every journal. It shouldn't have to read in the entirety of every journal file to answer a systemctl status $foo.service style query AIUI.

I've stopped all the services to not fuss with io readings of test now and:

* it took 3 seconds (shorter than before because there is nothing going in the system)
* it ran my hard drives up to 150MB/s entire time

Out of 1.3 GB of journal it read ~450MB of it. So yeah, while it isn't visiting *every* single byte of the file, it's still terrible

> My efforts historically mostly focused on the journald/writer side of things, not the consumer/reader side.

Well, the way it is structured is pretty much unfixable to be efficient at finding log for a single service.

**Comment by vcaputo** [comment]
Well that's still a lot different from reading *all* of the entries, and indicates to me that no, we don't have a silly bug here to go fix.

An important implementation detail to keep in mind is the journal-file's use of mmap, relying heavily on the kernel's async read-ahead/around to achieve some semblance of decent performance.

The kernel doesn't have any knowledge of the journal-file layout/access patterns, and the layout isn't optimized to try pack as much query-relevant data contiguously for any given kernel read-ahead/around.  So we pull in a lot of "chaff" in those async reads triggered by the "wheat" accesses.

If you tune the kernel to use smaller read-ahead/around sizes, you'll find the performance gets even worse, despite reading much less data from disk.  This is because the userspace code accessing the data via mmap becomes blocked more often waiting on the granular faults, in a "death by a thousand papercuts" fashion, by becoming more synchronous with the kernel handling the faults.

I wonder if `userfaultfd` could be exploited to implement a layout-aware readahead that knows what query is being performed *and* where the data resides in the relevant journal files.  Maybe if sd-journal could handle its own page faults, we could avoid reading so much "chaff", without having to significantly change either the on-disk layout or the mmap-reliant implementation.  Some food for thought and experimentation perhaps...

**Comment by XANi** [comment]
> Well that's still a lot different from reading all of the entries, and indicates to me that no, we don't have a silly bug here to go fix.

Yes, it is not "silly bug", it's just a fundamental fault with log format and way it is written and accessed. *just* having logfile-per-service would cut that by orders of magnitude. Hell, even hashing logfile names ("service x goes into hash(servicename).[0..3])") would cut it by orders of magnitude.

Check status of a single service is the most often use of for checking the logs and somehow happens to also be worst case scenario for this dreary design.

**Comment by AbsurdlySuspicious** [comment]
As another point of comparison: I need to keep a lot of logs (tens of GB), and I regularly need to extract logs within a 24-48 hour span from about 2-4 weeks ago. I had to set up syslog-ng, the functionality of which I didn't really need, but I had to use it because extracting logs from journald in the scenario described above can take almost half an hour. This setup created a lot of journal files, but they were on a fast nvme ssd. 

So in syslog-ng I set up mysql backend and text log per service with logrotate as backup. I haven't even had to use mysql log yet, because searching multiple rotated files of a single service log (to find the start date of the range I need to extract) with ripgrep takes several seconds (!) instead of tens of minutes. Given that logrotate rotates logs every week, and I keep them for at least a year, this is also a lot of files that kernel has no knowledge of to create efficient caches, let alone that ripgrep doesn't know that file lines are sorted by date, but it still performs a giant lot better.

**Comment by XANi** [comment]
I did some playing around  and results has been quite interesting:

The data journalctl returns:
```
# journalctl |dd of=/dev/zero bs=1M
0+1594 records in
0+1594 records out
6531134 bytes (6.5 MB, 6.2 MiB) copied, 0.615293 s, 10.6 MB/s
# journalctl |wc -l 
44464
```
6.5MB worth of text, 44k lines of text. It takes (freshly cleaned out before test, but same ratios as when it got to 1GB):
```
du -x /var/log/journal/ -h
57M	/var/log/journal/44cf6f547971fc33309d1e9e000002e7
```
now I assume binary log entries have a bit more metadata but I've seen anywhere betweeh ~850 to >1k per log entry when it got to 1GB allowed for /var/log/journal, which seems excessive.

So yeah, no wonder it is slow if each entry takes so much more than actual info in it. That on top of no way to lookup known

for comparision, just dumping it to SQLite database:
```
 sqlite /tmp/log.sqlite 'select count(*) from log'
51834
[14:55:57]cthulhu:/tmp☠ ls -s /tmp/log.sqlite -h
11M /tmp/log.sqlite

```
(there is more log entries because it is sucked off rsyslog that does some remote logging for me)
adding index bumps it to around 15MB. I assume adding boot_id would bump it a bit but not to level of journalctl.

**Comment by davispuh** [comment]
More log fun, whenever I try to query service status I get


```
# journalctl -u service --since -10m
Too many journal files (limit is at 7168) in scope, ignoring file '/var/log/journal/873a5f55f2aa4b33b2568baca40e6a91/system@13e94865c50347e3a62de9c85946c18e-0000000000991ecf-0006131335742f28.journal'.

-- No entries --
```

It took forever to even tell me "No entries" and I know it's lying, there for sure are some entries...

**Comment by Weltolk** [comment]
same issue

**Comment by dtardon** [comment]
> It took forever to even tell me "No entries" and I know it's lying, there for sure are some entries...

It also told you why it couldn't produce meaningful output, but you chose to ignore that part:
```
Too many journal files (limit is at 7168) in scope, ignoring file '/var/log/journal/873a5f55f2aa4b33b2568baca40e6a91/system@13e94865c50347e3a62de9c85946c18e-0000000000991ecf-0006131335742f28.journal'.
```
If you have a custom journald config. that produces that many files, you should rethink it. Running `journalctl --vacuum-time=X/--vacuum-size=Y/--vacuum-files=Z` may help for a time.

**Comment by XANi** [comment]
@dtardon why it can't just shard by app name ? Current approach is excessively wasteful in every aspect. It's equivalent of selecting entire database any time you need a record by one of mainly used columns...

**Comment by dtardon** [comment]
> why it can't just shard by app name ?

I don't understand what you mean by this...

> Current approach is excessively wasteful in every aspect. It's equivalent of selecting entire database any time you need a record by one of mainly used columns...

To get the last few records, it's not necessary to read all journal files _in full_. But _all_ journal files have to be opened to figure out where these last few records are. Which can be a problem if there are too many of them there, like in your case. (In fact, you have even more files than are allowed by the `sd-journal` library, so some--well, one--of them are not opened, which necessarily leads to gaps in the log.)

**Comment by XANi** [comment]
>>   why it can't just shard by app name ?

> I don't understand what you mean by this...

write to journal file named `hash(service_name)[0,2]_system@bootid` instead of `system@bootid`

So if hashed service name hashes to da39a3ee5e6b4b0d3255bfef95601890afd80709 the file name would be `da_system@bootid` and the tools would need to only open 1/256 of total data if it was looking for a particular service.

That's just ugly workaround but really anything is better than current garbage implementation.

Hell, just save a name last file service wrote log into together with rest of the service info. 

> To get the last few records, it's not necessary to read all journal files in full. But all journal files have to be opened to figure out where these last few records are. 

In case of last log entry being few days ago (or rotated out) that still ends up being majority of logs being read. For my 1GB of logs asking for app with rotated out logs does around 440MB of reads on simple `systemctl status <service>`

> Which can be a problem if there are too many of them there, like in your case.

Wrong person, I'm not the guy hitting 7k limit. I have only 1GB of journal and it still takes multiple seconds on my NAS, because it is a NAS so all free ram gets used to caching other stuff, so the moment I need to use any of the systemd commands touching logs there is none of older logs in disk cache and it takes forever

**Comment by dtardon** [comment]
> write to journal file named `hash(service_name)[0,2]_system@bootid` instead of `system@bootid`
> 
> So if hashed service name hashes to da39a3ee5e6b4b0d3255bfef95601890afd80709 the file name would be `da_system@bootid` and the tools would need to only open 1/256 of total data if it was looking for a particular service.

It would also produce 256 times more files, increasing disk usage (active journal files have a fixed minimal size--4 MB IIRC) and making hitting the file limit more probable. That's not mentioning that `journald` would have to keep 256 opened file descriptors instead of 1 (FDs are a limited resource).
 
> That's just ugly workaround but really anything is better than current garbage implementation.

There are more use cases for journal than showing a few lines of log in `systemctl status`, you know. You propose to (possibly) improve perf. of `systemctl status` at the cost of making everything else worse.

**Comment by Self-Perfection** [comment]
> To get the last few records, it's not necessary to read all journal files in full. But all journal files have to be opened to figure out where these last few records are.

All files has to be opened in absolutely worst case. In the most common scenario it should be enough to check most recent journal file and find that it has enough entries. And the most recent file can be found just by ordering them by modification time, it needs very small amount of IO.

Reading from all journal files for each status report is extremely inefficient.

**Comment by dtardon** [comment]
> > To get the last few records, it's not necessary to read all journal files in full. But all journal files have to be opened to figure out where these last few records are.
> 
> All files has to be opened in absolutely worst case. In the most common scenario it should be enough to check most recent journal file and find that it has enough entries. And the most recent file can be found just by ordering them by modification time, it needs very small amount of IO.

```
touch $(ls -tr /var/log/journal/$(</etc/machine-id)/system*.journal | head -n 1)
```
Suddenly the oldest archived journal file is the most recent... And that's just an artificial counter-example. There are completely legitimate time shifts (e.g., DST changes) that may cause the active journal file not to be _the most recent_ one. Journal still tries to preserve correct ordering of logs in such case, but it couldn't do that relying just on mtimes.

**Comment by Self-Perfection** [comment]
Well there are should be ways to optimize this. At least journalctl can safely assume that current active journal file is newer then all archived files. Thus if active journal file contains enough records to populate status output, then archived files should not be accessed.

> There are completely legitimate time shifts (e.g., DST changes) that may cause the active journal file not to be the most recent one. 
I think any time shifts affect credibility of both filesystem file timestamps and timestamps stored within files in roughly the same way.
If file timestamp seem unreliable, then let's store last record timestamp in filenames of archived files.

**Comment by chutzimir** [comment]
When reading logs I have the expectation to see them in the order they are produced. If there are time jumps during that time, I am OK with it.

I imagine old-school admins who are used to tailing /var/log/messages would expect the same thing.

> There are completely legitimate time shifts (e.g., DST changes) that may cause the active journal file not to be _the most recent_ one.

There is also a reasonable expectation that the system time is monotonic. I cannot think of a "legitimate" time shift? DST is certainly not it, because the timestamps (mtimes) are recorded as Unix times which is not affected by DST.

E.g., here is an example of DST moving back while Unix time is moving forward
```
$ for t in 2024-11-03T{04..07}:{00,30}Z; do t=$(date +%s -d $t); echo "Unix time: $t, $(TZ=UTC date -d @$t), $(TZ=America/New_York date -d @$t)"; done
Unix time: 1730606400, Sun  3 Nov 04:00:00 UTC 2024, Sun  3 Nov 00:00:00 EDT 2024
Unix time: 1730608200, Sun  3 Nov 04:30:00 UTC 2024, Sun  3 Nov 00:30:00 EDT 2024
Unix time: 1730610000, Sun  3 Nov 05:00:00 UTC 2024, Sun  3 Nov 01:00:00 EDT 2024
Unix time: 1730611800, Sun  3 Nov 05:30:00 UTC 2024, Sun  3 Nov 01:30:00 EDT 2024
Unix time: 1730613600, Sun  3 Nov 06:00:00 UTC 2024, Sun  3 Nov 01:00:00 EST 2024
Unix time: 1730615400, Sun  3 Nov 06:30:00 UTC 2024, Sun  3 Nov 01:30:00 EST 2024
Unix time: 1730617200, Sun  3 Nov 07:00:00 UTC 2024, Sun  3 Nov 02:00:00 EST 2024
Unix time: 1730619000, Sun  3 Nov 07:30:00 UTC 2024, Sun  3 Nov 02:30:00 EST 2024
```

**Comment by HLFH** [comment]
We forgot to celebrate the 10-year anniversary of this issue. Happy 10th birthday, issue #2460! 🎂

N.B.: If there’s anything we can do to help fix this issue, we’d be happy to help.

**Comment by XANi** [comment]
TBF it did get better
```
[10:38:10]cthulhu:~☠ echo 3 > /proc/sys/vm/drop_caches ; time systemctl status nginx 
● nginx.service - A high performance web server and a reverse proxy server
     Loaded: loaded (/usr/lib/systemd/system/nginx.service; enabled; preset: enabled)
     Active: active (running) since Thu 2026-03-26 12:05:46 CET; 22h ago
 Invocation: ed1eaec9288348da81e344ab0ec1d664
       Docs: man:nginx(8)
    Process: 2936 ExecStartPre=/usr/sbin/nginx -t -q -g daemon on; master_process on; (code=exited, status=0/SUCCESS)
    Process: 3140 ExecStart=/usr/sbin/nginx -g daemon on; master_process on; (code=exited, status=0/SUCCESS)
   Main PID: 3141 (nginx)
      Tasks: 5 (limit: 18997)
     Memory: 5.1M (peak: 8.1M)
        CPU: 7.758s
     CGroup: /system.slice/nginx.service
             ├─3141 "nginx: master process /usr/sbin/nginx -g daemon on; master_process on;"
             ├─3142 "nginx: worker process"
             ├─3143 "nginx: worker process"
             ├─3144 "nginx: worker process"
             └─3145 "nginx: worker process"

real	0m1,731s
user	0m0,017s
sys	0m0,052s
```
and the journald slowness can be skipped by `-n 0`

**Comment by Self-Perfection** [comment]
I have finally implemented a workaround: https://github.com/Self-Perfection/journald-archive

For now it is implemented as a Debian package.
TLDR: install package, enjoy faster `systemctl status`, to view all journald logs use `journalctl-all` cmd, simple `journalctl` will show only most recent logs.

-------


The idea is to split journald logs to small hot tier and cold storage archive. Once logs get rotated they moved to compressed btrfs filesystem. All systemd related commands (`systemctl status`, `journalctl` etc ) see only most recent log files and therefore _comparatively_ fast.

Package provides `journalctl-all` command to view combined journalctl logs. It might be slow if you have hundreds of journald files in archive! But i suspect overall it will be faster than querying the same amount of logs on uncompressed filesystem.
